### PR TITLE
Datadog exporter improvements

### PR DESCRIPTION
**Pull request by BrynCooke**
This PR updates the Datadog exporter with a backport from the latest code. 
The exporter code has been placed in the `apollo-router/src/plugins/telemetry/tracing/datadog_exporter` directory.

It also fixes:
* `span.kind` not being propagated
* Missing span stats. (Users may override configuration for these).

Span metrics are enabled for major built in spans by default.

```yaml
telemetry:
  exporters:
    tracing:
      experimental_response_trace_id:
        enabled: true
        header_name: apollo-custom-trace-id
        format: datadog
      common:
        service_name: router
      datadog:
        enabled: true
        batch_processor:
          scheduled_delay: 100ms
        span_metrics: # New config option to enable or disable span metrics. This is a free format span to boolean map.
          supergraph: false

  instrumentation:
    spans:
      mode: spec_compliant
      supergraph:
        attributes:
          graphql.operation.name: true
```

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [x] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

[ROUTER-363]: https://apollographql.atlassian.net/browse/ROUTER-363?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .changesets/fix_bryn_datadog_exporter_span_kind.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+### Datadog `span.kind` now populated ([PR #5609](https://github.com/apollographql/router/pull/5609))
+
+Datadog traces use `span.kind` to differentiate between different types of spans. 
+This change ensures that the `span.kind` is correctly populated using the Open Telemetry span kind which has a 1-2-1 mapping to thouse set out in [dd-trace](https://github.com/DataDog/dd-trace-go/blob/main/ddtrace/ext/span_kind.go).
```
Comment (garypen) [inline .changesets/fix_bryn_datadog_exporter_span_kind.md:4]:
```suggestion
This change ensures that the `span.kind` is correctly populated using the Open Telemetry span kind which has a 1-2-1 mapping to those set out in [dd-trace](https://github.com/DataDog/dd-trace-go/blob/main/ddtrace/ext/span_kind.go).
```

File: .changesets/fix_bryn_datadog_exporter_span_metrics.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+### Datadog span metrics now supported ([PR #5609](https://github.com/apollographql/router/pull/5609))
```
Comment (garypen) [inline .changesets/fix_bryn_datadog_exporter_span_metrics.md:1]:
```suggestion
### Datadog span metrics are now supported ([PR #5609](https://github.com/apollographql/router/pull/5609))
```

File: .changesets/fix_bryn_datadog_exporter_span_metrics.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+When using the APM view in Datadog, span metrics will be displayed for any span that was a top level span or has been explicitly has `_dd.measured` flag set.
```
Comment (garypen) [inline .changesets/fix_bryn_datadog_exporter_span_metrics.md:3]:
```suggestion
When using the APM view in Datadog, span metrics will be displayed for any span that was a top level span or has the `_dd.measured` flag set.
```

File: apollo-router/src/plugins/telemetry/consts.rs [hunk 4]
```diff
@@ -18,8 +18,9 @@ pub(crate) const REQUEST_SPAN_NAME: &str = "request";
 pub(crate) const QUERY_PLANNING_SPAN_NAME: &str = "query_planning";
 pub(crate) const HTTP_REQUEST_SPAN_NAME: &str = "http_request";
 pub(crate) const SUBGRAPH_REQUEST_SPAN_NAME: &str = "subgraph_request";
+pub(crate) const QUERY_PARSING_SPAN_NAME: &str = "parse_query";
```
Comment (garypen) [inline apollo-router/src/plugins/telemetry/consts.rs:21]: This looks like a "drive-by" fix. Should this be called out in the changelog as additional functionality or split into its own PR?

File: apollo-router/src/plugins/telemetry/tracing/datadog.rs [hunk 5]
```diff
@@ -79,6 +92,19 @@ pub(crate) struct Config {
     /// http_request -> http.route
     #[serde(default)]
     resource_mapping: HashMap<String, String>,
+
+    /// Which spans will be eligable for span stats to be collected for viewing in the APM view.
+    /// Defaults to true for `router` and `supergraph`, `subgraph`, `subgraph_request`, `http_request` and `query_planning`.
```
Comment (garypen) [inline apollo-router/src/plugins/telemetry/tracing/datadog.rs:97]: This comment isn't accurate. It is missing at least execution. I would expect the defaults to be as listed in `BUILT_IN_SPAN_NAMES`, so probably better to just reference that.

File: apollo-router/src/plugins/telemetry/tracing/datadog.rs [hunk 6]
```diff
@@ -170,13 +196,97 @@ impl TracingConfigurator for Config {
                     .expect("cargo version is set as a resource default;qed")
                     .to_string(),
             )
+            // This prevents spurious errors in the log when talking to the agent
```
Comment (garypen) [inline apollo-router/src/plugins/telemetry/tracing/datadog.rs:199]: Another "drive-by", which probably merits some mention in the changelog.

File: docs/source/configuration/telemetry/exporters/tracing/datadog.mdx [hunk 7]
```diff
@@ -166,7 +166,7 @@ The default resource mappings are:
 | `subgraph_request`      | `graphql.operation.name`    |
 | `http_request`          | `http.route`                |
 
-You may override these mappings by specifying the `resource_mapping` configuration:
+You may override these mappings by specifying `resource_mapping` configuration:
```
Comment (garypen) [inline docs/source/configuration/telemetry/exporters/tracing/datadog.mdx:169]:
```suggestion
You may override these mappings by specifying the `resource_mapping` configuration:
```

**Review by garypen — COMMENTED** [review]
Configuration comment: When we update otel to the latest library, what will be the impact on this configuration that we're adding. Will it continue to work unchanged or will it become redundant or ...?

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/plugins/telemetry/consts.rs (hunk 4 above)
Comment (abernix) [reply apollo-router/src/plugins/telemetry/consts.rs:21]: I guess it was https://github.com/apollographql/router/pull/5609/files#diff-cab56c50a115affb2a92de4523cc0b85f6c6b7f921693f8e7e5b4c3b34eb78e2L35 originally?  (in another file)

Comment (BrynCooke) [reply apollo-router/src/plugins/telemetry/consts.rs:21]: Yeah, I just moved this, it's not a rename or anything. It's not changelog worthy.

File: apollo-router/src/plugins/telemetry/tracing/datadog.rs (hunk 5 above)
Comment (BrynCooke) [reply apollo-router/src/plugins/telemetry/tracing/datadog.rs]: Fixed in 3ee8b3a. The reason not to reference `BUILT_IN_SPAN_NAMES` is that the docs are exposed to the user for via the json schema.

File: apollo-router/src/plugins/telemetry/tracing/datadog.rs (hunk 6 above)
Comment (BrynCooke) [reply apollo-router/src/plugins/telemetry/tracing/datadog.rs]: Removed and will pull into a separate PR.

Comment (BrynCooke) [reply apollo-router/src/plugins/telemetry/tracing/datadog.rs]: [0390461](https://github.com/apollographql/router/pull/5609/commits/0390461f5cfa4ffe6e86ca579ab66511211190e4)

**Review by garypen — APPROVED** [review]
